### PR TITLE
feat: UbiquityPoolFacet fuzz tests

### LINKED ISSUE #925: Fuzzing tests
We should implement fuzzing tests for [LibUbiquityPool](https://github.com/ubiquity/ubiquity-dollar/blob/a1b11e1b944227144d7cc19ab320220d9fce8b69/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol). I think it would be enough to fuzz only "user scenarios":
- [Dollar token mint](https://github.com/ubiquity/ubiquity-dollar/blob/a1b11e1b944227144d7cc19ab320220d9fce8b69/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol#L486)
- [Dollar token redeem](https://github.com/ubiquity/ubiquity-dollar/blob/a1b11e1b944227144d7cc19ab320220d9fce8b69/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol#L594)
- [collecting redemption](https://github.com/ubiquity/ubiquity-dollar/blob/a1b11e1b944227144d7cc19ab320220d9fce8b69/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol#L710)

Possible solutions for fuzzing tests (I would simply start with foundry since we're using it as a development framework + it also contains invariant testing):
- https://book.getfoundry.sh/forge/fuzz-testing
- https://github.com/crytic/echidna
- https://github.com/crytic/medusa

What should be done:
1. Impelent fuzzing tests for the scenarios described above (you may create them in a separate test file like `UbiquityPoolFacet.fuzz.t.sol` so we could distinguish unit and fuzz tests)
2. Setup testing CI this way:
a) On PR fuzzing tests should run with a small number of runs (so that we could merge PRs fast)
b) On merge to the `development` branch fuzzing tests should run with a great number of runs to test many cases (keep in mind that github action runnners can run for 6 hours)

### PR DESCRIPTION
**Pull request by gitcoindev**
Resolves https://github.com/ubiquity/ubiquity-dollar/issues/925

Skeleton based on UbiquityPoolFacet tests.

### CODE REVIEW

**Comment by gitcoindev** [comment]
This pull request is still a draft. It contains a skeleton for fuzz tests, I will keep updating and hopefully make as ready for review during or after the weekend.

**Comment by gitcoindev** [comment]
I am playing with the fuzzer a lot and discovered an interesting case. When redeem threshold is set to $0.99 , the curve price must be larger than 990000999999999999 due to conversion between 6 digit and 18 digit curve prices in `getDollarPriceUsd() <= poolStorage.redeemPriceThreshold` comparison, i.e

```
vm.assume(dollarPriceUsd > 990000999999999999); // 0.99e18 , greater than redeem threshold
```

I was stuck for some time with the failing fuzz test but after debugging found the edge value. Fuzzer is great, as setting mocked price to 990000999999999998 it almost immediately finds this and reports the case.

**Comment by gitcoindev** [comment]
I will finish with the whole test suite till Friday EOD and set the pr as ready for review.

**Comment by rndquu** [comment]
> I am playing with the fuzzer a lot and discovered an interesting case. When redeem threshold is set to $0.99 , the curve price must be larger than 990000999999999999 due to conversion between 6 digit and 18 digit curve prices in `getDollarPriceUsd() <= poolStorage.redeemPriceThreshold` comparison, i.e
> 
> ```
> vm.assume(dollarPriceUsd > 990000999999999999); // 0.99e18 , greater than redeem threshold
> ```
> 
> I was stuck for some time with the failing fuzz test but after debugging found the edge value. Fuzzer is great, as setting mocked price to 990000999999999998 it almost immediately finds this and reports the case.

This seems to be a low severity issue but keeping in mind that nobody from the Sherlock audit have reported this bug it seems that nobody fuzzed the `UbiquityPoolFacet` contract extensively.

File: packages/contracts/test/diamond/facets/UbiquityPoolFacet.fuzz.t.sol [hunk 1]
```diff
@@ -240,47 +240,42 @@ contract UbiquityPoolFacetFuzzTest is DiamondTestSetup {
...
-
         vm.prank(user);
         vm.expectRevert("Governance slippage");
         ubiquityPoolFacet.mintDollar(
             0, // collateral index
             100e18, // Dollar amount
-            90e18, // min amount of Dollars to mint
-            10e18, // max collateral to send
-            0, // max Governance tokens to send
+            98e18, // min amount of Dollars to mint (2% fee included)
```
Comment (0x4007) [inline packages/contracts/test/diamond/facets/UbiquityPoolFacet.fuzz.t.sol]: 2% seems steep? Why was this decided to be the default? Please tag me in your response since these days I'm not on GitHub much and receive a lot of notifications.

Comment (rndquu) [reply packages/contracts/test/diamond/facets/UbiquityPoolFacet.fuzz.t.sol]: @0x4007 

That 2% fee is used solely in the test files.

Mint and redeem fees were [set](https://github.com/ubiquity/ubiquity-dollar/blob/e4c1fae5ef5e9ba40bcb84c671f2ba39f0ab5208/packages/contracts/migrations/development/Deploy001_Diamond_Dollar_Governance.s.sol#L316-L317) to 0% on deployment + they are adjustable.

**Comment by gitcoindev** [comment]
I pushed the remaining fuzz tests also for redemption delay, setting as ready for review now.

File: packages/contracts/test/diamond/facets/UbiquityPoolFacet.fuzz.t.sol (hunk 1 above)
Comment (gitcoindev) [reply packages/contracts/test/diamond/facets/UbiquityPoolFacet.fuzz.t.sol]: @0x4007 yes, as @rndquu mentioned, on deployment fees are set to 0, can be adjusted at any time by the admin.

File: packages/contracts/foundry.toml [hunk 2]
```diff
@@ -42,3 +42,4 @@ src = 'src/dollar'
 
 [profile.intense.fuzz]
 runs = 100000
+max_test_rejects = 90000
```
Comment (rndquu) [inline packages/contracts/foundry.toml]: @gitcoindev How did you come up with the `90000` number? By default it is [set](https://book.getfoundry.sh/reference/config/testing#max_test_rejects) to `65536`. Why do we need to increase it to this exact number?

**Comment by rndquu** [comment]
@gitcoindev 

1. Could you add 2 more fuzz tests:
a) If user mints X Dollars he gets exactly X Dollars (minus fee)
b) If user redeems X Dollars exactly X Dollars are subtracted from user balance
2. Pls notice the original issue spec also describe setting up CI

Overall the PR looks great, we will iterate on the fuzzing tests later

File: packages/contracts/foundry.toml (hunk 2 above)
Comment (gitcoindev) [reply packages/contracts/foundry.toml]: @rndquu when I executed longer fuzzing sessions I frequently encountered the error about max rejects number. It happens when vm.assume call rejects more than 65536 input values during test case execution. I increased the default number to 90k to make it sufficient for 100k fuzz iterations. I realized that we do not have to fuzz for 6 hours as the fuzzer is quite good in finding edge cases almost immediately, 100k input values should be more than enough for most tests.

**Comment by gitcoindev** [comment]
> @gitcoindev
> 
> 1. Could you add 2 more fuzz tests:
>    a) If user mints X Dollars he gets exactly X Dollars (minus fee)
>    b) If user redeems X Dollars exactly X Dollars are subtracted from user balance
> 2. Pls notice the original issue spec also describe setting up CI
> 
> Overall the PR looks great, we will iterate on the fuzzing tests later

Sure, thank you for the feedback.

**Comment by gitcoindev** [comment]
Hi @rndquu @0x4007 @molecula451 the PR is ready to review, I executed QA against my fork: https://github.com/gitcoindev/ubiquity-dollar/actions/runs/9534134896/job/26278201422 . The CI setup is ready and according to the original spec. Fuzz tests run by default with 256 inputs on PR commits and the 'deep fuzz' workflow with 100000 inputs https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/deep-fuzz.yml is executed after merge to the development branch.

**Comment by molecula451** [comment]
hi @gitcoindev can we instead organize the tests inside a "fuzz" folder,

`packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol` or something similar
 
i think this is our initial sketching/drawing for fuzzing inside the repo, besides the SMTChecker we've done before, but untouching the .sol files, this is the first so it would be nice the addition of the folder for future development

**Comment by gitcoindev** [comment]
@molecula451 I moved fuzz tests as requested, could you please have a look once again now?

**Comment by molecula451** [comment]
the foundry standard library allow us to use `assume()` as an abstraction to meet certain conditions on functionality, to fuzz input and get a corresponding value according to the data

we have moved the files and structured a new way for proper fuzzing tests file in the future

this PR meets the correct condition at issue #925 implementing

Minting: https://github.com/ubiquity/ubiquity-dollar/blob/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol#L154

Reedeeming:  https://github.com/ubiquity/ubiquity-dollar/blob/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol#L333

of the corresponding 

user https://github.com/ubiquity/ubiquity-dollar/blob/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol#L32 an assuming arbitrary input, therefore i am approving this PR

**Review by molecula451 — APPROVED** [review]
the foundry standard library allow us to use `assume()` as an abstraction to meet certain conditions on functionality, to fuzz input and get a corresponding value according to the data

we have [moved](https://github.com/ubiquity/ubiquity-dollar/tree/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz) the files and structured a new way for proper fuzzing tests file in the future

this PR meets the correct condition at issue #925 implementing

- Correct foundry.toml profile for fuzzing selection

Minting: https://github.com/ubiquity/ubiquity-dollar/blob/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol#L154

Reedeeming:  https://github.com/ubiquity/ubiquity-dollar/blob/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol#L333

of the corresponding user https://github.com/ubiquity/ubiquity-dollar/blob/7dce7be237666ed7c62a1d0baae2bec66ebe5ac1/packages/contracts/test/fuzz/diamond/facets/UbiquityPoolFacet.fuzz.t.sol#L32 and assuming arbitrary input, therefore i am approving this PR

**Review by rndquu — APPROVED** [review]

**Comment by rndquu** [comment]
@gitcoindev Did you experiment with how long the deep fuzz [workflow](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/.github/workflows/deep-fuzz.yml) runs for, let's say, `1_000_000` [runs](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/packages/contracts/foundry.toml#L44) ? I want to find out a value so that deep fuzz workflow could run for the amount of hours close to github max CI limit (6 hours).

**Comment by 0x4007** [comment]
> @gitcoindev Did you experiment with how long the deep fuzz [workflow](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/.github/workflows/deep-fuzz.yml) runs for, let's say, `1_000_000` [runs](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/packages/contracts/foundry.toml#L44) ? I want to find out a value so that deep fuzz workflow could run for the amount of hours close to github max CI limit (6 hours).

Can try a binary search and just push a few commits.

**Comment by molecula451** [comment]
i think in part we're going to see the runs with new upcoming PRs and improve the addition workflow from there

**Comment by gitcoindev** [comment]
> > @gitcoindev Did you experiment with how long the deep fuzz [workflow](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/.github/workflows/deep-fuzz.yml) runs for, let's say, `1_000_000` [runs](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/packages/contracts/foundry.toml#L44) ? I want to find out a value so that deep fuzz workflow could run for the amount of hours close to github max CI limit (6 hours).
> 
> Can try a binary search and just push a few commits.

During my experiments and tests implementation I figured out that 100 000 was more than enough to find all edge cases for this Ubiquity pool lib. Before pushing the commit I modified e.g. vm.assume, expected values or lib implementation, with very small values or large values and fuzzer always found the problem with less than 10 000 iterations. I set deep fuzz to 100 000 to be sure that it will find new issues, but of course we can experiment and increase to full 6 hrs and check if it brings any value.

**Comment by rndquu** [comment]
> > > @gitcoindev Did you experiment with how long the deep fuzz [workflow](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/.github/workflows/deep-fuzz.yml) runs for, let's say, `1_000_000` [runs](https://github.com/ubiquity/ubiquity-dollar/blob/64ab8b9e378abe7cc019924bd43e7891d6d9d884/packages/contracts/foundry.toml#L44) ? I want to find out a value so that deep fuzz workflow could run for the amount of hours close to github max CI limit (6 hours).
> > 
> > 
> > Can try a binary search and just push a few commits.
> 
> During my experiments and tests implementation I figured out that 100 000 was more than enough to find all edge cases for this Ubiquity pool lib. Before pushing the commit I modified e.g. vm.assume, expected values or lib implementation, with very small values or large values and fuzzer always found the problem with less than 10 000 iterations. I set deep fuzz to 100 000 to be sure that it will find new issues, but of course we can experiment and increase to full 6 hrs and check if it brings any value.

https://github.com/ubiquity/ubiquity-dollar/issues/943
